File: src/main/java/com/oycm/algorithm/ba/opposite/Solution_17.scala
package com.oycm.algorithm.ba.opposite

object Solution_17 {

  /**
   * 1577. 数的平方等于两数乘积的方法数
   * https://leetcode.cn/problems/number-of-ways-where-square-of-number-is-equal-to-product-of-two-numbers/description
   *
   * 两个整数数组 nums1 和 nums2
   *
   * 根据以下规则形成的三元组的数目（类型 1 和类型 2 ）
   *  - 类型 1：三元组 (i, j, k) ，如果 nums1[i] pow 2 == nums2[j] * nums2[k] 其中 0 <= i < nums1.length 且 0 <= j < k < nums2.length
   *  - 类型 2：三元组 (i, j, k) ，如果 nums2[i] pow 2 == nums1[j] * nums1[k] 其中 0 <= i < nums2.length 且 0 <= j < k < nums1.length
   *
   * @param nums1 正整数数组
   * @param nums2 正整数数组
   * @return
   */
  def numTriplets(nums1: Array[Int], nums2: Array[Int]): Int = {
    /*
    哈希表计算：先用 map 记录 两个数组平方的数量记为 pow1 pow2
    再计算 两个数组各自乘积的数量，几位 p1 p2
    遍历 pow1 的额 key 取 p2 对于的 value 相乘

    时间复杂度 O(n^2) 计算乘积的次数
    空间复杂度 O(n^2)
    */
    def pow(nums: Array[Int]): scala.collection.mutable.Map[Long, Int] = {
      val map = scala.collection.mutable.Map[Long, Int]()
      for (num <- nums) {
        // 注意计算精度丢失问题
        val temp = Math.pow(num, 2).toLong
        map(temp) = map.getOrElse(temp, 0) + 1
      }
      map
    }

    def product(nums: Array[Int]): scala.collection.mutable.Map[Long, Int] = {
      val map = scala.collection.mutable.Map[Long, Int]()
      val n = nums.length
      for (i <- 0 until n - 1) {
        for (j <- i + 1 until n) {
          val temp = nums(i).toLong * nums(j)
          map(temp) = map.getOrElse(temp, 0) + 1
        }
      }
      map
    }

    def calculate(pow: scala.collection.mutable.Map[Long, Int], product: scala.collection.mutable.Map[Long, Int]): Int = {
      var ans = 0
      for ((k, v) <- pow) {
        if (product.contains(k)) {
          ans += v * product(k)
        }
      }
      ans
    }

    val pow1 = pow(nums1)
    val pow2 = pow(nums2)
    val p1 = product(nums1)
    val p2 = product(nums2)
    calculate(pow1, p2) + calculate(pow2, p1)
  }

  def main(args: Array[String]): Unit = {
    println(numTriplets(Array(7, 4), Array(5, 2, 8, 9)) == 1)
    println(numTriplets(Array(1, 1), Array(1, 1, 1)) == 9)
    println(numTriplets(Array(7, 7, 8, 3), Array(1, 2, 9, 7)) == 2)
    println(numTriplets(Array(4, 7, 9, 11, 23), Array(3, 5, 1024, 12, 18)) == 0)
    println(numTriplets(Array(43024, 99908), Array(1864)) == 0)
    println(numTriplets(Array(100000, 100000, 100000, 100000), Array(100000, 100000, 100000, 100000)) == 48)
  }
}

File: src/main/java/com/oycm/hot100/ordinary_array/Solution_17.java
package com.oycm.hot100.ordinary_array;

public class Solution_17 {

    /**
     * 41. <a href="https://leetcode.cn/problems/first-missing-positive/description/">缺失的第一个正数</a>
     *
     * @param nums 未排序的整数数组 nums[i] [-2^31, 2^31-1]
     * @return 找出其中没有出现的最小的正整数
     */
    public int firstMissingPositive(int[] nums) {
        /*
        题解思路：先思考暴力的做法
        O(n) 空间，hash 表记录 nums 所有元素，遍历 [1, n + 1]，如果哪个数不存在，哪个就是缺失的第一个正整数
        O(1) 空间，O(n^2) 时间，遍历 [1, n + 1] 看 nums 中是否存在该数
        对于一个长为 n 的数组，第一个最小的正整数，范围一定在 [1, n+1] 中。
            如果 [1, n] 都在数组中出现，答案就是 n+1；
            否则，答案就是 [1, n] 中的最小正整数。
        数组的长度恰好为 n，是否能设计一个 hash 表，表示整数在 [1, n] 中存在。
        遍历 nums[i] 如果 num = nums[i] <= n，对数组进行打标记，逻辑如下：
            nums[num-1] = -abs(nums[num-1]) 用负数符号标记这个数存在，前面的判断也需要假设绝对值处理；注意，后续再访问，负号一直存在
            这样打完标记后，第一个整数下标 + 1 就是答案
        因为 nums 中存在负数，所以需要对其进行预处理，将负数转换成不在 [1, n] 范围的正整数
         */
        int n = nums.length;
        // 预处理小于等于 0 的数
        for (int i = 0; i < n; i++) {
            if (nums[i] <= 0) {
                nums[i] = n + 1;
            }
        }
        for (int i = 0; i < n; i++) {
            int num = Math.abs(nums[i]);
            if (num <= n) {
                // 打标记
                nums[num - 1] = -Math.abs(nums[num - 1]);
            }
        }
        for (int i = 0; i < n; i++) {
            if (nums[i] > 0) {
                return i + 1;
            }
        }
        return n + 1;
    }

    public int firstMissingPositive_swap(int[] nums) {
        /*
        题解思路：交换位置
        记 x = nums[i]，把 x 交换到正确的位置。
            如果 x > 0 && x <= n，temp = nums[x-1]，把 nums[x-1] = x, x = temp 继续保持这个逻辑交换，直到 nums[x-1] == x
         */
        int n = nums.length;
        for (int i = 0; i < n; i++) {
            while (0 < nums[i] && nums[i] <= n && nums[nums[i] - 1] != nums[i]) {
                int j = nums[i] - 1;
                // 交换，把 nums[i] 交换到正确的位置
                int temp = nums[j];
                nums[j] = nums[i];
                nums[i] = temp;
            }
        }
        for (int i = 0; i < n; i++) {
            if (i + 1 != nums[i]) {
                return i + 1;
            }
        }
        return n + 1;
    }

}
